File: sdk/src/main/java/com/checkout/sdk/CheckoutClient.java
package com.checkout.sdk;

import android.content.Context;

import com.android.volley.VolleyError;
import com.checkout.sdk.core.TokenResult;
import com.checkout.sdk.request.CardTokenizationRequest;
import com.checkout.sdk.request.GooglePayTokenisationRequest;
import com.checkout.sdk.response.CardTokenizationFail;
import com.checkout.sdk.response.CardTokenizationResponse;
import com.checkout.sdk.response.GooglePayTokenisationFail;
import com.checkout.sdk.response.GooglePayTokenisationResponse;
import com.checkout.sdk.utils.Environment;
import com.checkout.sdk.utils.HttpUtils;
import com.google.gson.Gson;

import org.json.JSONException;
import org.json.JSONObject;

public class CheckoutClient {

    public String key;

    public interface TokenCallback {
        void onTokenResult(TokenResult tokenResult);
    }

    /**
     * This is interface used as a callback for when the card token is generated
     */
    public interface OnTokenGenerated {
        void onTokenGenerated(CardTokenizationResponse response);

        void onError(CardTokenizationFail error);

        void onNetworkError(VolleyError error);
    }

    /**
     * This is interface used as a callback for when the google pay token is generated
     */
    public interface OnGooglePayTokenGenerated {
        void onTokenGenerated(GooglePayTokenisationResponse response);

        void onError(GooglePayTokenisationFail error);

        void onNetworkError(VolleyError error);
    }

    private Context mContext;
    private final Environment mEnvironment;
    private final CheckoutClient.TokenCallback tokenCallback;
    private CheckoutClient.OnGooglePayTokenGenerated mGooglePayTokenListener;


    public CheckoutClient(Context context, String key, Environment environment, CheckoutClient.TokenCallback tokenCallback) {
        this.mContext = context;
        this.key = key;
        this.mEnvironment = environment;
        this.tokenCallback = tokenCallback;
    }

    public CheckoutClient.TokenCallback getTokenCallback() {
        return tokenCallback;
    }

    public Environment getEnvironment() {
        return mEnvironment;
    }

    /**
     * This method is used to generate a card token.
     * <p>
     * It takes a {@link CardTokenizationRequest} as the argument and it will perform a
     * HTTP Post request to generate the token. it is important to you select an environment and
     * provide your public key before calling tis method. Moreover it is important to set a callback
     * {@link CheckoutClient.OnTokenGenerated} so you can receive the token back.
     * <p>
     * If you are using the UI of the SDK this method will be called automatically, but you still
     * need to provide the callback, key and environment when initialising this class
     *
     * @param request Custom request body to be used in the HTTP call.
     */
    public void generateToken(CardTokenizationRequest request, OnTokenGenerated tokenListener) {

        // Initialise the HTTP utility class
        HttpUtils http = new HttpUtils(mContext);

        // Provide a callback for when the token request is completed
        http.setTokenListener(tokenListener);

        // Using Gson to convert the custom request object into a JSON string for use in the HTTP call
        Gson gson = new Gson();
        String jsonBody = gson.toJson(request);

        try {
            http.generateToken(key, mEnvironment.getTokenHost() + "/" + Environment.TOKEN_PATH, jsonBody);
        } catch (JSONException e) {
            e.printStackTrace();
        }
    }

    /**
     * This method used to create a token that can be used in a server environment to create a
     * charge. It will take a GooglePay payload in JSON string format. The payload is usually generated in
     * the handlePaymentSuccess method shown in the Google Pay example from Google (token.getToken())
     *
     * @param payload Google Pay Payload
     */
    public void generateGooglePayToken(String payload) throws JSONException {

        JSONObject googlePayToken = new JSONObject(payload);

        // Initialise the HTTP utility class
        HttpUtils http = new HttpUtils(mContext);

        // Provide a callback for when the token request is completed
        http.setGooglePayTokenListener(mGooglePayTokenListener);

        GooglePayTokenisationRequest gPay = new GooglePayTokenisationRequest();

        gPay
                .setSignature(googlePayToken.getString("signature"))
                .setProtocolVersion(googlePayToken.getString("protocolVersion"))
                .setSignedMessage(googlePayToken.getString("signedMessage"));

        // Using Gson to convert the custom request object into a JSON string for use in the HTTP call
        Gson gson = new Gson();
        String jsonBody = gson.toJson(gPay);

        try {
            String url = mEnvironment.getGooglePayHost() + "/" + Environment.GOOGLE_PAY_PATH;
            http.generateGooglePayToken(key, url, jsonBody);
        } catch (JSONException e) {
            e.printStackTrace();
        }

    }

    /**
     * This method used to set a callback for Google Pay handling.
     *
     * @return CheckoutClient to allow method chaining
     */
    public CheckoutClient setGooglePayListener(OnGooglePayTokenGenerated listener) {
        this.mGooglePayTokenListener = listener;
        return this;
    }
}

File: sdk/src/main/java/com/checkout/sdk/CheckoutClient.kt
package com.checkout.sdk

import android.content.Context
import com.checkout.sdk.api.ApiFactory
import com.checkout.sdk.api.TokenApi
import com.checkout.sdk.core.TokenResult
import com.checkout.sdk.executors.Coroutines
import com.checkout.sdk.request.CardTokenizationRequest
import com.checkout.sdk.request.GooglePayTokenizationRequest
import com.checkout.sdk.request.TokenRequest
import com.checkout.sdk.response.CardTokenizationFail
import com.checkout.sdk.response.GooglePayTokenisationFail
import com.checkout.sdk.response.TokenFail
import com.checkout.sdk.response.TokenResponse
import com.checkout.sdk.utils.Environment
import com.google.gson.Gson
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import okhttp3.ResponseBody


class CheckoutClient(
    private val key: String,
    private val tokenApi: TokenApi,
    private val coroutines: Coroutines,
    private val tokenCallback: TokenCallback,
    var progressCallback: ProgressCallback? = null
) {

    /**
     * Request a token:
     * - either straight from card details: CardTokenizationRequest, or
     * - from a GooglePay generated token: GooglePayTokenizationRequest
     */
    fun requestToken(request: TokenRequest) {
        val deferred = when (request) {
            is CardTokenizationRequest -> tokenApi.getTokenAsync(key, request)
            is GooglePayTokenizationRequest -> tokenApi.getGooglePayTokenAsync(key, request)
            else -> throw IllegalArgumentException("Unknown class: ${request.javaClass}")
        }
        coroutines.IOScope.launch {
            val response =
                try {
                    deferred.await()
                } catch (e: Exception) {
                    handleRequestError(e)
                    return@launch
                }

            if (response.isSuccessful) {
                handleResponseSuccessful(response.body()!!)
            } else {
                handleResponseFailure(response.errorBody()!!, request.javaClass)
            }
        }
        progressCallback?.onProgressChanged(true)
    }

    private suspend fun handleRequestError(e: Exception) {
        withContext(coroutines.Main) {
            tokenCallback.onTokenResult(TokenResult.TokenResultNetworkError(e))
            progressCallback?.onProgressChanged(false)
        }
    }

    private suspend fun handleResponseSuccessful(body: TokenResponse) {
        withContext(coroutines.Main) {
            tokenCallback.onTokenResult(TokenResult.TokenResultSuccess(body))
            progressCallback?.onProgressChanged(false)
        }
    }

    private suspend fun handleResponseFailure(
        errorBody: ResponseBody,
        clazz: Class<TokenRequest>
    ) {
        val tokenError = getTokenError(errorBody.string(), clazz)
        withContext(coroutines.Main) {
            tokenCallback.onTokenResult(
                TokenResult.TokenResultTokenizationFail(
                    tokenError
                )
            )
            progressCallback?.onProgressChanged(false)
        }
    }

    private fun getTokenError(errorString: String, clazz: Class<TokenRequest>): TokenFail {
        val fail = when (clazz) {
            CardTokenizationRequest::class.java -> CardTokenizationFail::class.java
            GooglePayTokenizationRequest::class.java -> GooglePayTokenisationFail::class.java
            else -> throw IllegalArgumentException("Unknown class: $clazz")
        }
        return Gson().fromJson(errorString, fail)
    }

    interface ProgressCallback {
        fun onProgressChanged(inProgress: Boolean)
    }

    companion object {
        /**
         * Convenience method for simple creation of a CheckoutClient
         *
         * @param context The android context (activity or application context)
         * @param key Your public key (https://sandbox.checkout.com/settings/channels -> Public Key)
         * @param environment Sandbox (for testing) or live
         * @param tokenCallback The callback for receiveing your Token, or an Error
         */
        fun create(context: Context, key: String, environment: Environment, tokenCallback: TokenCallback): CheckoutClient {
            val apiFactory = ApiFactory(context, environment)
            val tokenApi = apiFactory.api
            return CheckoutClient(
                key,
                tokenApi,
                Coroutines(),
                tokenCallback
            )
        }
    }

    /**
     * Receive your result from your token request (requestToken) via this callback
     *
     * TokenResult will be one of the following:
     * - TokenResultSuccess if you receive the Token successfully
     * - TokenResultTokenizationFail if you there is an error creating your token (typically due
     *   to incorrect values being sent or a server error)
     * - TokenResultNetworkError if Network problems (e.g. being Offline) case a failure
     */
    interface TokenCallback {
        fun onTokenResult(tokenResult: TokenResult)
    }
}
